File: frontend/src/types/review.ts
import type { Book } from './book'
import type { User } from './auth'

export interface Review {
  id: number
  title: string
  content: string
  score: number
  rating: number | null
  status: string
  created_at: string
  book: Book
  user: User
}

export interface ReviewCreate {
  book_id: number
  title: string
  content: string
}

export interface SearchResults {
  books: import('./book').Book[]
  reviews: Review[]
  users: { username: string; role: string }[]
}

export interface Notification {
  id: number
  type: 'comment' | 'vote' | 'follow'
  read: boolean
  created_at: string
  actor_username: string
  review_id: number | null
  review_title: string | null
}

File: backend/models/review.py
from sqlalchemy import Integer, String, Text, DateTime, ForeignKey, func
from sqlalchemy.orm import Mapped, mapped_column, relationship
from database import Base


class Review(Base):
    __tablename__ = "review"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    user_id: Mapped[int] = mapped_column(Integer, ForeignKey("user.id"), nullable=False)
    book_id: Mapped[int] = mapped_column(Integer, ForeignKey("book.id"), nullable=False)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    score: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    rating: Mapped[int | None] = mapped_column(Integer, nullable=True)
    status: Mapped[str] = mapped_column(String(10), default="active", nullable=False)
    created_at: Mapped[DateTime] = mapped_column(DateTime, server_default=func.now())

    user: Mapped["User"] = relationship("User", back_populates="reviews")
    book: Mapped["Book"] = relationship("Book", back_populates="reviews")
    comments: Mapped[list["Comment"]] = relationship("Comment", back_populates="review")
    votes: Mapped[list["Vote"]] = relationship("Vote", back_populates="review")
